src/main/haskell/kore/src/Kore/ASTUtils/SmartPatterns.hs

```diff
@@ -70,9 +70,10 @@ pattern Ceil_
     -> PureMLPattern level var
 
 pattern DV_
-  :: Sort Object
+  :: () => (level ~ Object) =>
```

Comment:
Why the `() =>`?

Reply:
Pattern synonyms take *two* contexts: the first is the required context, the second is the provided context. (Or, precondition and postcondition if you prefer.) This says: nothing is required to attempt matching this pattern, and if it does match then you may assume `level ~ Object`. The prior version of `DV_` had `level` replaced by `Object` explicitly; that is equivalent to having the constraints `(level ~ Object) => (level ~ Object) => ...`. In other words, the old pattern synonym required you to know the pattern was an object-level pattern before attempting to match `DV_`.  The [GHC manual](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#typing-of-pattern-synonyms) describes all the details of typing pattern synonyms if you want to know more.

Reply:
Thanks for the explanation! I removed the `=> ()` part and it seemed to compile, so I thought I'd ask.